### PR TITLE
8319778: Remove unreachable code in ObjectSynchronizer::exit

### PR DESCRIPTION
**Pull request by xmas92**
In `ObjectSynchronizer::exit` when unlocking on an inflated monitor with `LM_LIGHTWEIGHT` there is a check if the owner is anonymous. That branch should never be take and would indicate that something is wrong with either `ObjectSynchronizer::inflate` or that `ObjectSynchronizer::exit` is trying to unlock a monitor not owned by the `current` thread.

Changes this to an assert instead. Did not condition the assert on `LM_LIGHTWEIGHT` as this property should hold regardless of locking mode.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319778](https://bugs.openjdk.org/browse/JDK-8319778): Remove unreachable code in ObjectSynchronizer::exit (**Enhancement** - P4)

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**)
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16602/head:pull/16602` \
`$ git checkout pull/16602`

Update a local copy of the PR: \
`$ git checkout pull/16602` \
`$ git pull https://git.openjdk.org/jdk.git pull/16602/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16602`

View PR using the GUI difftool: \
`$ git pr show -t 16602`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16602.diff">https://git.openjdk.org/jdk/pull/16602.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16602#issuecomment-1805630489)

### CODE REVIEW

**Comment by xmas92** [comment]
## Testing
  - `linux-x64`, `linux-x64-debug`
    - `LM_LEGACY`
      - [X] tier 1-5
    - `LM_LIGHTWEIGHT`
      - [X] tier 1-5
  - [X] GitHub actions

**Review by rkennke — APPROVED** [review]
Looks good!

**Review by dholmes-ora — COMMENTED** [review]
Sorry I don't follow the rationale here. If the object is fast-locked to T1 and T2 tries to lock, it will inflate and have to install the anonymous-owner. WhenT1 exits it will see this state.

**Comment by xmas92** [comment]
> Sorry I don't follow the rationale here. If the object is fast-locked to T1 and T2 tries to lock, it will inflate and have to install the anonymous-owner. WhenT1 exits it will see this state.

This is true. But the current post condition enforced by `ObjectSynchronizer::inflate` is that if the `current` JavaThread owns the monitor then it also fixes the the owner (, and in the recursive lightweight implementation, the recursions) on the `ObjectMonitor`. So in `ObjectSynchronizer::exit` which may only ever be called on an object owned by `current` this must be the case. 

If it would call exit on an object not owned by the current thread it may or may not say anonymous. If it were to say anonymous then the `current` JavaThread's lock stack would not contain the object, and it would not fix the owner. It would instead `assert(false, "Non-balanced monitor enter/exit!")` inside `ObjectMonitor::exit`.

**Review by dholmes-ora — APPROVED** [review]
Thanks - I had missed that `inflate()` already executed the same logic.

File: src/hotspot/share/runtime/synchronizer.cpp [hunk 1]
```diff
@@ -638,13 +638,7 @@ void ObjectSynchronizer::exit(oop object, BasicLock* lock, JavaThread* current)
...
   // dropped inside exit() and the ObjectMonitor* must be !is_busy().
   ObjectMonitor* monitor = inflate(current, object, inflate_cause_vm_internal);
-  if (LockingMode == LM_LIGHTWEIGHT && monitor->is_owner_anonymous()) {
-    // It must be owned by us. Pop lock object from lock stack.
-    LockStack& lock_stack = current->lock_stack();
-    oop popped = lock_stack.pop();
-    assert(popped == object, "must be owned by this thread");
-    monitor->set_owner_from_anonymous(current);
-  }
+  assert(!monitor->is_owner_anonymous(), "must not be");
```
Comment (dcubed-ojdk) [inline src/hotspot/share/runtime/synchronizer.cpp:641]: Agreed that the block from old L641 -> L647 is impossible after the call
to `inflate()` on L640. Thanks for adding that assert.

The other reason is if the owner came in here with a lightweight monitor,
then that would have been handled via L571 -> L592 and we never should
have reached this block.

**Review by dcubed-ojdk — APPROVED** [review]
Thumbs up.

Please run Tier1 testing prior to integration if that has not yet been done.
I don't expect any issues on non-Linux platforms, but you never know...

**Comment by xmas92** [comment]
Ran the extra tier 1 tests on all platforms. Thanks for the reviews.

/integrate

**Pull request closed without merging**